### PR TITLE
Altered accounts by block nonce/shard

### PR DESCRIPTION
**Pull request by mariusmihaic**
- Added supported for altered accounts by block `nonce/hash`, which forwards the request from proxy to specified shard block/nonce. More details about the endpoint can be found [here](https://github.com/ElrondNetwork/elrond-go/pull/4538)
- Added endpoints:
```
<proxy>/block/<shard id>/altered-accounts/by-nonce/<nonce>
<proxy>/block/<shard id>/altered-accounts/by-hash/<hash>
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,6 +71,14 @@ For more details, go to [docs.elrond.com](https://docs.elrond.com/sdk-and-tools/
...
+- `/v1.0/block/:shardID/altered-accounts/by-nonce/:nonce?tokens=token1,token2`    (GET) --> returns altered accounts in the given block by nonce, filtered out by given tokens
+- `/v1.0/block/:shardID/altered-accounts/by-nonce/:nonce?tokens=token1,token2&withMetadata=true`    (GET) --> returns altered accounts in the given block by nonce, filtered out by given tokens. For each esdt, it will also add its metadata
+- `/v1.0/block/:shardID/altered-accounts/by-hash/:hash`    (GET) --> returns altered accounts in the given block by hash
+- `/v1.0/block/:shardID/altered-accounts/by-hash/:hash?tokens=token1,token2`    (GET) --> returns altered accounts in the given block by hash, filtered out by given tokens
+- `/v1.0/block/:shardID/altered-accounts/by-hash/:hash?tokens=token1,token2&withMetadata=true`    (GET) --> returns altered accounts in the given block by hash, filtered out by given tokens. For each esdt, it will also add its metadata
+
+Please note that in order for `altered-accounts` endpoint to work with historical balances, observers should have this flag **on** when started.
```
Comment (bogdan-rosianu) [inline README.md:81]: rephrase:
```
Please note that `altered-accounts` endpoints will only work if the backing observers of the Proxy have support for historical balances (`--operation-mode historical-balances` when starting the node)
```

File: api/groups/baseBlockGroup.go [hunk 2]
```diff
@@ -118,3 +118,83 @@ func (group *blockGroup) byNonceHandler(c *gin.Context) {
 
 	c.JSON(http.StatusOK, blockByNonceResponse)
 }
+
+func (group *blockGroup) alteredAccountsByNonceHandler(c *gin.Context) {
+	shardID, err := shared.FetchShardIDFromRequest(c)
+	if err != nil {
+		shared.RespondWith(
```
Comment (bogdan-rosianu) [inline api/groups/baseBlockGroup.go:125]: respondwithbadrequest?

File: api/groups/baseBlockGroup.go [hunk 3]
```diff
@@ -118,3 +118,83 @@ func (group *blockGroup) byNonceHandler(c *gin.Context) {
...
+			nil,
+			apiErrors.ErrCannotParseShardID.Error(),
+			data.ReturnCodeRequestError,
+		)
+		return
+	}
+
+	nonce, err := shared.FetchNonceFromRequest(c)
+	if err != nil {
+		shared.RespondWith(
```
Comment (bogdan-rosianu) [inline api/groups/baseBlockGroup.go:137]: respondwithbadrequest?

File: api/groups/baseBlockGroup.go [hunk 4]
```diff
@@ -118,3 +118,83 @@ func (group *blockGroup) byNonceHandler(c *gin.Context) {
...
+		return
+	}
+
+	c.JSON(http.StatusOK, blockByNonceResponse)
+}
+
+func (group *blockGroup) alteredAccountsByHashHandler(c *gin.Context) {
+	shardID, err := shared.FetchShardIDFromRequest(c)
+	if err != nil {
+		shared.RespondWith(
```
Comment (bogdan-rosianu) [inline api/groups/baseBlockGroup.go:165]: respondwithbadrequest?

File: api/groups/baseBlockGroup.go [hunk 5]
```diff
@@ -118,3 +118,83 @@ func (group *blockGroup) byNonceHandler(c *gin.Context) {
...
+			nil,
+			apiErrors.ErrCannotParseShardID.Error(),
+			data.ReturnCodeRequestError,
+		)
+		return
+	}
+
+	hash, err := shared.FetchHashFromRequest(c)
+	if err != nil {
+		shared.RespondWith(
```
Comment (bogdan-rosianu) [inline api/groups/baseBlockGroup.go:177]: respondwithbadrequest?

File: go.mod [hunk 6]
```diff
@@ -3,24 +3,26 @@ module github.com/ElrondNetwork/elrond-proxy-go
 go 1.17
 
 require (
-	github.com/ElrondNetwork/elastic-indexer-go v1.2.30
-	github.com/ElrondNetwork/elrond-go v1.3.27
-	github.com/ElrondNetwork/elrond-go-core v1.1.18
-	github.com/ElrondNetwork/elrond-go-crypto v1.0.1
-	github.com/ElrondNetwork/elrond-go-logger v1.0.7
+	github.com/ElrondNetwork/elastic-indexer-go v1.2.44-0.20221006131929-a7c3dec53ba1
+	github.com/ElrondNetwork/elrond-go v1.3.45-0.20221020085243-978bf0f08a13
```
Comment (bogdan-rosianu) [inline go.mod:7]: should we use proper tags before merging?

File: process/blockProcessor.go [hunk 7]
```diff
@@ -343,3 +346,53 @@ func (bp *BlockProcessor) GetInternalStartOfEpochMetaBlock(epoch uint32, format
...
+	for _, observer := range observers {
+		var response data.AlteredAccountsApiResponse
+
+		_, err := bp.proc.CallGetRestEndPoint(observer.Address, path, &response)
+		if err != nil {
+			log.Error("altered accounts request", "observer", observer.Address, "error", err.Error())
+			continue
+		}
+
+		log.Info("altered accounts request", "shard id", observer.ShardId, "nonce", nonce, "observer", observer.Address)
```
Comment (sstanculeanu) [inline process/blockProcessor.go:367]:
```suggestion
		log.Info("altered accounts request by nonce", "shard id", observer.ShardId, "nonce", nonce, "observer", observer.Address)
```

File: process/blockProcessor.go [hunk 8]
```diff
@@ -343,3 +346,53 @@ func (bp *BlockProcessor) GetInternalStartOfEpochMetaBlock(epoch uint32, format
...
+		return nil, err
+	}
+	path := common.BuildUrlWithAlteredAccountsQueryOptions(fmt.Sprintf("%s/%d", alteredAccountByBlockNonce, nonce), options)
+
+	for _, observer := range observers {
+		var response data.AlteredAccountsApiResponse
+
+		_, err := bp.proc.CallGetRestEndPoint(observer.Address, path, &response)
+		if err != nil {
+			log.Error("altered accounts request", "observer", observer.Address, "error", err.Error())
```
Comment (sstanculeanu) [inline process/blockProcessor.go:363]:
```suggestion
			log.Error("altered accounts request by nonce", "observer", observer.Address, "error", err.Error(), "nonce", nonce)
```

File: process/blockProcessor.go [hunk 9]
```diff
@@ -343,3 +346,53 @@ func (bp *BlockProcessor) GetInternalStartOfEpochMetaBlock(epoch uint32, format
...
+		return nil, err
+	}
+	path := common.BuildUrlWithAlteredAccountsQueryOptions(fmt.Sprintf("%s/%s", alteredAccountByBlockHash, hash), options)
+
+	for _, observer := range observers {
+		var response data.AlteredAccountsApiResponse
+
+		_, err := bp.proc.CallGetRestEndPoint(observer.Address, path, &response)
+		if err != nil {
+			log.Error("altered accounts request", "observer", observer.Address, "error", err.Error())
```
Comment (sstanculeanu) [inline process/blockProcessor.go:388]:
```suggestion
			log.Error("altered accounts request by hash", "observer", observer.Address, "error", err.Error(), "hash", hash)
```

File: process/blockProcessor.go [hunk 10]
```diff
@@ -343,3 +346,53 @@ func (bp *BlockProcessor) GetInternalStartOfEpochMetaBlock(epoch uint32, format
...
+			continue
+		}
+
+		log.Info("altered accounts request", "shard id", observer.ShardId, "hash", hash, "observer", observer.Address)
```
Comment (sstanculeanu) [inline process/blockProcessor.go:392]:
```suggestion
		log.Info("altered accounts request by hash", "shard id", observer.ShardId, "hash", hash, "observer", observer.Address)
```

File: api/groups/baseBlockGroup.go [hunk 11]
```diff
@@ -118,3 +118,83 @@ func (group *blockGroup) byNonceHandler(c *gin.Context) {
...
+
+	options, err := parseAlteredAccountOptions(c)
+	if err != nil {
+		shared.RespondWithValidationError(c, apiErrors.ErrBadUrlParams, err)
+		return
+	}
+
+	blockByNonceResponse, err := group.facade.GetAlteredAccountsByNonce(shardID, nonce, options)
+	if err != nil {
+		shared.RespondWith(c, http.StatusInternalServerError, nil, err.Error(), data.ReturnCodeInternalError)
```
Comment (sstanculeanu) [inline api/groups/baseBlockGroup.go:155]: `shared.RespondWithInternalError`

File: README.md (hunk 1 above)
Comment (mariusmihaic) [reply README.md:81]: Right, very good suggestion, thanks

File: api/groups/baseBlockGroup.go (hunk 2 above)
Comment (mariusmihaic) [reply api/groups/baseBlockGroup.go:125]: Right, better this way

File: api/groups/baseBlockGroup.go (hunk 3 above)
Comment (mariusmihaic) [reply api/groups/baseBlockGroup.go:137]: Right, refactored

File: api/groups/baseBlockGroup.go (hunk 5 above)
Comment (mariusmihaic) [reply api/groups/baseBlockGroup.go:177]: Right, refactored

File: go.mod (hunk 6 above)
Comment (mariusmihaic) [reply go.mod:7]: These are the commit hashes from feat branches on elrond-go/go-core. I don't think there is any tag for that

File: process/blockProcessor.go (hunk 7 above)
Comment (mariusmihaic) [reply process/blockProcessor.go:367]: Right, added it

File: process/blockProcessor.go (hunk 8 above)
Comment (mariusmihaic) [reply process/blockProcessor.go:363]: Right, added it

File: process/blockProcessor.go (hunk 9 above)
Comment (mariusmihaic) [reply process/blockProcessor.go:388]: Right, added it

File: process/blockProcessor.go (hunk 10 above)
Comment (mariusmihaic) [reply process/blockProcessor.go:392]: Added it

File: api/groups/baseBlockGroup.go (hunk 11 above)
Comment (mariusmihaic) [reply api/groups/baseBlockGroup.go:155]: I have no outer and inner error to pass as asrguments for that func, so I would leave it as it is

**Review by sstanculeanu — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**